adam_core/backend/pyoorb.py

```diff
@@ -0,0 +1,326 @@
+            Epochs converted into the PYOORB format.
+        """
+        num_times = len(epochs)
+        if time_scale == "UTC":
+            time_scale_list = [1 for i in range(num_times)]
+        elif time_scale == "UT1":
+            time_scale_list = [2 for i in range(num_times)]
+        elif time_scale == "TT":
+            time_scale_list = [3 for i in range(num_times)]
+        elif time_scale == "TAI":
+            time_scale_list = [4 for i in range(num_times)]
+        else:
+            raise ValueError("time_scale should be one of {'UTC', 'UT1', 'TT', 'TAI'}")
+
+        epochs_pyoorb = np.array(
+            list(np.vstack([epochs, time_scale_list]).T), dtype=np.double, order="F"
+        )
+        return epochs_pyoorb
+
+    def _propagate_orbits(self, orbits: Orbits, times: Time) -> Orbits:
+        """
+        Propagate orbits using PYOORB.
+
+        Parameters
+        ----------
+        orbits : `~thor.orbits.orbits.Orbits` (N)
+            Orbits to propagate.
+        times : `~astropy.time.core.Time` (M)
+            Times to which to propagate orbits.
+
+        Returns
+        -------
+        propagated : `~thor.orbits.orbits.Orbits` (N * M)
+            Orbits propagated to each time in times.
+        """
+        # Convert orbits into PYOORB format
+        orbits_pyoorb = self._configure_orbits(
+            orbits.cartesian.values.filled(),
+            orbits.cartesian.times.tt.mjd,
+            "cartesian",
+            "TT",
+            magnitude=None,
+            slope=None,
+        )
+
+        # Convert epochs into PYOORB format
+        epochs_pyoorb = self._configure_epochs(times.tt.mjd, "TT")
+
+        # Propagate orbits to each epoch and append to list
+        # of new states
+        states_list = []
+        orbits_pyoorb_i = orbits_pyoorb.copy()
+        for epoch in epochs_pyoorb:
+            orbits_pyoorb_i, err = oo.pyoorb.oorb_propagation(
+                in_orbits=orbits_pyoorb_i,
+                in_epoch=epoch,
+                in_dynmodel=self.dynamical_model,
+            )
+            states_list.append(orbits_pyoorb_i)
+
+        # Convert list of new states into a pandas data frame
+        # These states at the moment will always be return as cartesian
+        # state vectors
+        # elements = ["x", "y", "z", "vx", "vy", "vz"]
+        # Other PYOORB state vector representations:
+        # "keplerian":
+        #    elements = ["a", "e", "i", "Omega", "omega", "M0"]
+        # "cometary":
+        #    elements = ["q", "e", "i", "Omega", "omega", "T0"]
+        states = np.concatenate(states_list)
+
+        # Extract cartesian states from PYOORB results
+        orbit_ids_ = states[:, 0].astype(int)
+        x = states[:, 1]
+        y = states[:, 2]
+        z = states[:, 3]
+        vx = states[:, 4]
+        vy = states[:, 5]
+        vz = states[:, 6]
+        mjd_tt = states[:, 8]
+
+        # Convert output epochs to TDB
```

Comment:
This seems a bit surprising - why always convert to tdb?

Reply:
This is to match what JPL does: by default: orbits in TDB, observations in UTC